docs/_includes/Statements.md

```diff
@@ -239,6 +240,242 @@ Note that the form
 
 is diagnosed as a label in which the user forgot the **label** keyword.
 
+## Update with Failure Statement (`:-`)
+````grammar
+UpdateFailureStmt  =
+    [ Lhs { "," Lhs } ]
+    ":-"
+    [ "expect" ]
+    Expression(allowLemma: false, allowLambda: false) { "," Rhs } 
+````
+
+An update statement using `:-` instead of `:=` exits execution of a method immediately 
+if a failure is reported. 
+This is a language feature somewhat analogous to exceptions in other languages.
+
+To use this form of update,
+
+ * the caller must have a first out-parameter whose type matches the output of `PropagateFailure` applied to the first output of the callee
+ * if the RHS of the update-with-failure statement is a method call, the first out-parameter of the callee must be failure-compatible
+ * if instead the RHS of the update-with-failure statement is one or more expressions, the first of these expressions must be a value with a failure-compatible type
+ * if the failure-compatible type of the RHS does not have an Extract method,
+then the LHS of the `:-` statement has one less expression than the RHS 
+(or than the number of out-parameters from the method call)
+ * if the failure-compatible type of the RHS does have an Extract method,
+then the LHS of the `:-` statement has the same numnber of expressions as the RHS 
+(or than the number of out-parameters from the method call)
+and the type of the first LHS expression must be assignable from the return type of the Extract method
+* the IsFailure, PropagateFailure and Extract methods may not be ghost
+
+A failure-compatible type is a type that has the following (each with no in-parameters and one out-parameter):
+
+ * a method or function `IsFailure()` that returns a `bool`
+ * a method or function `PropagateFailure()` that returns a value assignable to the first out-parameter of the caller
+ * an optional method or function `Extract()`
+
+### Failure compatible types
+
```

Comment:
Move this sub-sub-section earlier, and put the definition of failure-compatible types into this sub-sub-section.